.github/workflows/release.yml

```diff
@@ -1,98 +1,165 @@
       run: |
-        JAR_DOWNLOAD_URL=https://downloads.metabase.com/${{ github.ref_name }}/metabase.jar
-        if [[ ${{ github.ref_name }} == v1* ]]; then
-          JAR_DOWNLOAD_URL=https://downloads.metabase.com/enterprise/${{ github.ref_name }}/metabase.jar
-        fi
-        echo $JAR_DOWNLOAD_URL > url.txt
-        echo "----- Downloading Uberjar from $JAR_DOWNLOAD_URL -----"
-        curl -OL $JAR_DOWNLOAD_URL
-        stat ./metabase.jar
-        date | tee timestamp
+        echo "The version format is invalid!"
+        echo "It must start with either 'v0.' or 'v1.'."
+        echo "Please, try again."
+        exit 1
+    - name: find_release_artifact
+      id: find_release_artifact
+      uses: actions/github-script@v6
+      with:
+        result-encoding: string
+        script: |
+          const artifacts = await github.rest.actions.listArtifactsForRepo({
+            owner: context.repo.owner,
+            repo: context.repo.repo,
+            name: `metabase-${{ matrix.edition }}-${{ inputs.commit }}-uberjar`,
+            per_page: 1,
+          });
+          if (!artifacts.data?.artifacts?.[0]?.archive_download_url) {
+            throw new Error(`No artifacts found for ${{ inputs.commit }}`);
+          }
+          const download_url = artifacts.data.artifacts[0].archive_download_url;
+          console.log('download url', download_url);
+          return download_url;
+    - name: download-uberjar-artifact
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      run: |
+        curl -L \
+            -H "Accept: application/vnd.github+json" \
+            -H "Authorization: Bearer $GITHUB_TOKEN" \
+            -H "X-GitHub-Api-Version: 2022-11-28" \
+            ${{ steps.find_release_artifact.outputs.result }} \
+            -o mb.zip
+    - name: unzip uberjar artifact
+      run: unzip mb.zip
     - name: Verify that this is a valid JAR file
       run: file --mime-type ./metabase.jar | grep "application/zip"
     - name: Reveal its version.properties
       run: jar xf metabase.jar version.properties && cat version.properties
-    - name: Calculate SHA256 checksum
-      run: sha256sum ./metabase.jar | tee SHA256.sum
+    - name: Check JAR version properties
+      run: |
+        grep -q $(sha256sum ./metabase.jar) SHA256.sum && echo "checksums match" || (echo "checksum mismatch" && exit 1)
+        grep -q ${{ inputs.commit }} COMMIT-ID && echo "commit ids match" || (echo "commit-id mismatch" && exit 1)
+        grep -q $(cut -c 1-7 <<< ${{ inputs.commit }}) version.properties && echo "version.properties commit matches" || (echo "version.properties commit mismatch" && exit 1)
+        grep -q $(cut -c 3-99 <<< ${{ inputs.version }}) version.properties && echo "version.properties version number matches" || (echo "version.properties version number mismatch" && exit 1)
+        grep -q "SNAPSHOT" version.properties && (echo "jar is a snapshot" && exit 1) || echo "jar is not a snapshot build"
```

Comment:
This does a ton of redundant checks to make sure the JAR we downloaded matches the user inputs -  effectively making the version.properties in the jar the source of truth, and failing if anything doesn't match.